Mathlib/RingTheory/Grassmannian.lean

```diff
@@ -0,0 +1,67 @@
+/-
+Copyright (c) 2025 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import Mathlib.RingTheory.Spectrum.Prime.FreeLocus
+
+/-!
+# Grassmannians
+
+## Main definitions
+
+- `Grassmannian`: `G(k, M; R)` is the `k`ᵗʰ Grassmannian of the `R`-module `M`. It is defined to be
+  the set of submodules of `M` whose quotient is locally free of rank `k`. Note that this indexing
+  is the opposite of some indexing in literature, where this rank would be `n - k` instead, where
+  `M = R ^ n`.
+
+## TODO
+- Define the functor `Grassmannian.functor R M k` that sends an `R`-algebra `A` to the set
+  `G(k, A ⊗[R] M; A)`.
+- Define `chart x` indexed by `x : Fin k → M` as a subtype consisting of those
+  `N ∈ G(k, A ⊗[R] M; A)` such that the composition `R^k → M → M⧸N` is an isomorphism.
+- Define `chartFunctor x` to turn `chart x` into a subfunctor of `Grassmannian.functor`. This will
+  correspond to an affine open chart in the Grassmannian.
+- Grassmannians for schemes and quasi-coherent sheaf of modules.
+- Representability of `Grassmannian.functor R M k`.
+-/
+
+universe u v w
+
+open Module
+
+variable (R : Type u) [CommRing R] (M : Type v) [AddCommGroup M] [Module R M] (k : ℕ)
+
+/-- `G(k, M; R)` is the `k`ᵗʰ Grassmannian of the `R`-module `M`. It is defined to be the set of
+submodules of `M` whose quotient is locally free of rank `k`. Note that this indexing is the
+opposite of some indexing in literature, where this rank would be `n - k` instead, where
+`M = R ^ n`. -/
+def Grassmannian : Type v :=
+  { N : Submodule R M // Module.Finite R (M ⧸ N) ∧ Projective R (M ⧸ N) ∧
+    ∀ p, rankAtStalk (R := R) (M ⧸ N) p = k }
```

Comment:
It's not clear to me that using `Subtype` is helping you here. You could consider instead
```suggestion
structure Grassmannian extends Submodule R M where
  finite_quotient : Module.Finite R (M ⧸ toSubmodule)
  projective_quotient : Projective R (M ⧸ toSubmodule)
  rankAtStalk_eq : ∀ p, rankAtStalk (R := R) (M ⧸ toSubmodule) p = k
```